### PR TITLE
fix lowranceusr4 writer issues

### LINKED ISSUE #375: segmentation fault when writing USR ver 4 file
When trying to convert a GPX-file to lowrance usr file ver 4, I get a segmentation fault:

```
(xenial)erik@localhost:~/ gpsbabel -D 2 -i gpx -f plotter/to-plotter.gpx -o lowranceusr,wversion=4 -F to-plotter.usr

GPSBabel Version: 1.6.0 
main: Compiled with Qt 5.11.1 for architecture x86_64-little_endian-lp64
main: Running with Qt 5.11.1 on Ubuntu 16.04.6 LTS, x86_64
main: QLocale::system() is en_US
main: QLocale() is en_US
main: QTextCodec::codecForLocale() is UTF-8, mib 106
options: module/option=value: gpx/snlen="32" (=default)
options: module/option=value: gpx/elevprec="3" (=default)
options: module/option=value: lowranceusr/wversion="4"
options: module/option=value: lowranceusr/title="" (=default)
options: module/option=value: lowranceusr/serialnum="0" (=default)
options: module/option=value: lowranceusr/description="" (=default)
cet_util: Converting from "UTF-8" to "US-ASCII", done.
Lowrance USR data_write: Num Waypoints = 28
Lowrance USR data_write: Title = 'GPSBabel generated USR data file'
Lowrance USR data_write: Description = 'Waypoints, routes, and trails'
Lowrance USR adding waypt 001 (001) to table at index 0
Lowrance USR adding waypt Rpt001 (Rpt001) to table at index 1
Lowrance USR adding waypt Rpt004 (Rpt004) to table at index 2
Lowrance USR adding waypt Rpt005 (Rpt005) to table at index 3
Lowrance USR adding waypt 003 (003) to table at index 4
Lowrance USR adding waypt 004 (004) to table at index 5
Lowrance USR adding waypt NF2018-1 (NF2018-1) to table at index 6
Lowrance USR adding waypt NF2018-2 (NF2018-2) to table at index 7
Lowrance USR adding waypt NF2018-3 (NF2018-3) to table at index 8
Lowrance USR adding waypt NF2018-4 (NF2018-4) to table at index 9
Lowrance USR adding waypt NF2018-5 (NF2018-5) to table at index 10
Lowrance USR adding waypt NF2018-6 (NF2018-6) to table at index 11
Lowrance USR adding waypt NF2018-7 (NF2018-7) to table at index 12
Lowrance USR adding waypt NF2018-8 (NF2018-8) to table at index 13
Lowrance USR adding waypt NF2018-9 (NF2018-9) to table at index 14
Lowrance USR adding waypt NF2018-10 (NF2018-10) to table at index 15
Lowrance USR adding waypt NF2018-11 (NF2018-11) to table at index 16
Lowrance USR adding waypt NF2018-12 (NF2018-12) to table at index 17
Lowrance USR adding waypt NF2018-13 (NF2018-13) to table at index 18
Lowrance USR adding waypt NF2018-14 (NF2018-14) to table at index 19
Lowrance USR adding waypt NF2018-15 (NF2018-15) to table at index 20
Lowrance USR adding waypt NF2018-16 (NF2018-16) to table at index 21
Lowrance USR adding waypt NF2018-17 (NF2018-17) to table at index 22
Lowrance USR adding waypt NF2018-18 (NF2018-18) to table at index 23
Lowrance USR adding waypt WP0030-000 (WP0030-000) to table at index 24
Lowrance USR adding waypt WP0030-001 (WP0030-001) to table at index 25
Lowrance USR adding waypt WP0030-002 (WP0030-002) to table at index 26
Lowrance USR adding waypt WP0030-003 (WP0030-003) to table at index 27
Lowrance USR writing 28 waypoints
Lowrance USR writing out waypt 0 (001 - 001)
Lowrance USR writing out waypt 1 (Rpt001 - Rpt001)
Lowrance USR writing out waypt 2 (Rpt004 - Rpt004)
Lowrance USR writing out waypt 3 (Rpt005 - Rpt005)
Lowrance USR writing out waypt 4 (003 - 003)
Lowrance USR writing out waypt 5 (004 - 004)
Lowrance USR writing out waypt 6 (NF2018-1 - NF2018-1)
Lowrance USR writing out waypt 7 (NF2018-2 - NF2018-2)
Lowrance USR writing out waypt 8 (NF2018-3 - NF2018-3)
Lowrance USR writing out waypt 9 (NF2018-4 - NF2018-4)
Lowrance USR writing out waypt 10 (NF2018-5 - NF2018-5)
Lowrance USR writing out waypt 11 (NF2018-6 - NF2018-6)
Lowrance USR writing out waypt 12 (NF2018-7 - NF2018-7)
Lowrance USR writing out waypt 13 (NF2018-8 - NF2018-8)
Lowrance USR writing out waypt 14 (NF2018-9 - NF2018-9)
Lowrance USR writing out waypt 15 (NF2018-10 - NF2018-10)
Lowrance USR writing out waypt 16 (NF2018-11 - NF2018-11)
Lowrance USR writing out waypt 17 (NF2018-12 - NF2018-12)
Lowrance USR writing out waypt 18 (NF2018-13 - NF2018-13)
Lowrance USR writing out waypt 19 (NF2018-14 - NF2018-14)
Lowrance USR writing out waypt 20 (NF2018-15 - NF2018-15)
Lowrance USR writing out waypt 21 (NF2018-16 - NF2018-16)
Lowrance USR writing out waypt 22 (NF2018-17 - NF2018-17)
Lowrance USR writing out waypt 23 (NF2018-18 - NF2018-18)
Lowrance USR writing out waypt 24 (WP0030-000 - WP0030-000)
Lowrance USR writing out waypt 25 (WP0030-001 - WP0030-001)
Lowrance USR writing out waypt 26 (WP0030-002 - WP0030-002)
Lowrance USR writing out waypt 27 (WP0030-003 - WP0030-003)
Lowrance USR data_write: Num routes = 2
Lowrance USR writing route #0 (Route 002) with 4 waypts
Segmentation fault (core dumped)
```

### PR DESCRIPTION
**Pull request by tsteven4**
This should resolve #375.

In addition, further writer issues with lowranceusr version 4 should be resolved:

- BOMs were erroneously included in internal UTF-16LE strings.
- Timestamps were erroneously offset by 12 hours.
- The written number of mysterious attribute bytes didn't match the written count.

Also, rounding is now used when converting real latitude and longitude numbers to the integer format used by all lowranceusr files.  This makes round trip testing much more likely to succeed.

### CODE REVIEW

**Comment by tsteven4** [comment]
@bjkowalski , can you please review these changes?
@erkswede, can you test your conversion related to #375 on your device, and make sure the resulting version 4 file is valid?

Thanks

**Comment by bjkowalski** [comment]
Just got back into the country.  Will review in a few days.

**Comment by erkswede** [comment]
Sure, should I
"git clone https://github.com/tsteven4/gpsbabel.git"
and compile as usual?

**Comment by erkswede** [comment]
Ok I think I figured it out, did:
`git clone -b issue375 --single-branch https://github.com/tsteven4/gpsbabel.git`
and compiled. Needed to install gperf (wasn't needed in main branch).
Everything seems to work fine, no segmentation fault and the file can be imported into my B&G Vulcan 7 (non FS). This is the first file touched by something else than the Vulcan itself that I managed to import (USR ver 3 wasn't compatible for instance) and I only tested one file real quick. So don't rely to much on this test.
Thanks a bunch for your quick respons on this! Now I have a viable workflow from Plotter -> OpenCPN -> Plotter.

File: testo.d/lowranceusr.test [hunk 1]
```diff
@@ -68,22 +68,22 @@ rm -f ${TMPDIR}/lowrance*
...
 gpsbabel -i geo -f ${REFERENCE}/../geocaching.loc -o lowranceusr,wversion=2 -F ${TMPDIR}/lowrance1.usr
-bincompare ${TMPDIR}/lowrance1.usr ${REFERENCE}/lowrance.usr
+bincompare ${REFERENCE}/lowrance.usr ${TMPDIR}/lowrance1.usr
 
 # Use the file just created to see if can write back the same data
-gpsbabel -i lowranceusr -f ${TMPDIR}/lowrance1.usr -o lowranceusr,wversion=2 -F ${TMPDIR}/lowrance1.usr
-# Unfortunately precision issues cause mismatch on Lat/Long conversion, actual test validation commented out
-# bincompare ${REFERENCE}/lowrance.usr  ${TMPDIR}/lowrance1.usr
+gpsbabel -i lowranceusr -f ${TMPDIR}/lowrance1.usr -o lowranceusr,wversion=2 -F ${TMPDIR}/lowrance2.usr
+# At a minimum unkown altitude not making the round trip, actual test validation commented out
```
Comment (GPSBabelDeveloper) [inline testo.d/lowranceusr.test]: unknown

**Review by GPSBabelDeveloper — APPROVED** [review]
I'll leave the final approval to the author of the new code here.
Overall, this looks pretty sane. The fs_find_chain stuff really does need a c++ rethink some day. Maybe the format specific stuff becomes subclasses of points, but we have cases (at least in theory) where we could have multiple format specific data attached. Maybe we just add an array of QVariants (std::variant) of name/value pairs and get rid of all the pointer abuse.
I'll think on that a bit, but this seems like a bunch of good fixes to me...

**Comment by tsteven4** [comment]
I have the "bin compare ${TMPDIR}/lowrance-v4.gpx ${TMPDIR}/lowrance-v4~usr.gpx" test working and committed below.

The test "bincompare ${REFERENCE}/lowrance.usr ${TMPDIR}/lowrance2.usr" has issues with altitude.
There is one obvious code bug, at line 839 wpt_tmp->altitude has not been written, the constructor will
have set it to unknown_altitude = -99999999.0.  After the conversion to feet the else statement at 839 is always true.
I also question the whole stream version 0 stores altitude as float idea.  It doesn't seem to be true in our test cases.  @bjkowalski can you weigh in on this as well.  thanks. 
https://github.com/gpsbabel/gpsbabel/blob/2c02294c38fb63c4c35dd0b6015cb2f6dee578d3/lowranceusr.cc#L827-L844

File: testo.d/lowranceusr.test (hunk 1 above)
Comment (tsteven4) [reply testo.d/lowranceusr.test]: spelling fixed.

**Comment by bjkowalski** [comment]
Taken me much longer to get to this than anticipated.  I reviewed the changes and they all look good.  I need to do some more digging into the float concerns with stream version 0.

**Review by bjkowalski — APPROVED** [review]

**Comment by tsteven4** [comment]
@bjkowalski Thanks for the review.  I moved the stream version 0 altitude concerns to #399, if you can figure that out that would be great.

**Pull request merged**